File: 0752-open-the-lock/0752-open-the-lock.py
class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        start = "0000"
        deadends_set = set(deadends)

        if start in deadends_set:
            return -1

        queue = deque([ (start, 0) ])
        
        while queue:
            current, moves = queue.popleft()

            if current == target:
                return moves
            
            for i, slot in enumerate(current):
                new_up = current[:i] + str((int(slot) + 1) % 10) + current[i+1:]
                new_down = current[:i] + str((int(slot) - 1) % 10) + current[i+1:]

                if new_up not in deadends_set:
                    deadends_set.add(new_up)
                    queue.append((new_up, moves + 1))

                if new_down not in deadends_set:
                    deadends_set.add(new_down)
                    queue.append((new_down, moves + 1))

        return -1

File: 0752-open-the-lock/0752-open-the-lock.js
/**
 * @param {string[]} deadends
 * @param {string} target
 * @return {number}
 */
var openLock = function(deadends, target) {
    let visited = new Set( deadends );
    if( visited.has(target) ||  visited.has("0000") )
        return -1;
    
    let queue = [ [ [0, 0, 0, 0], 0 ] ] // start turns
    
    while( queue.length ) {
        let [ curr, turns ] = queue.shift();
        
        if( curr.join("") === target )
            return turns
        
        for( let opt of options(curr) ) {
            if( visited.has(opt.join("")) )
                continue;
            
            visited.add(opt.join(""));
            queue.push( [ opt, turns+1 ] );
        }
    }
    return -1;
    
    function options(curr) {
        let res = [];
        for( let i=0; i < curr.length; i++ ) {
            let curr1 = [ ...curr ];
            let temp = curr1[i];
            curr1[i] = (curr1[i] + 1)%10;
            res.push( [...curr1]);
            
            curr1[i] = temp;
            curr1[i] = (curr1[i] - 1+10)%10;
            res.push( [...curr1]);
        }
        return res;
    }
};